### PR TITLE
experimental <processorpath> compile mojo parameter

### PR DESCRIPTION
**Pull request by ifedorenko**
Signed-off-by: Igor Fedorenko <ifedorenko@salesforce.com>

### CODE REVIEW

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/jdt/CompilerJdt.java [hunk 1]
```diff
@@ -684,14 +684,13 @@ public boolean setSourcepath(List<File> dependencies, Set<File> sourceRoots) thr
...
-    } else if (processorpath.isEmpty()) {
-      this.processorpath = Collections.emptyList();
     } else {
-      throw new IllegalArgumentException();
+      this.processorpath = ImmutableList.copyOf(processorpath);
     }
     if (!isProcNone() && processorpathDigester.digestProcessorpath(this.processorpath)) {
       log.debug("Annotation processor path changed, recompiling all sources");
       strategy.enqueueAllSources();
+      strategy = new FullCompilationStrategy();
```
Comment (rexhoffman) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/jdt/CompilerJdt.java]: why is this line needed?  You just enqueued all sources

File: takari-lifecycle-plugin/src/test/java/io/takari/maven/plugins/compile/SimpleReactorReader.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.artifact.DefaultArtifact;
+import org.eclipse.aether.repository.WorkspaceRepository;
+import org.eclipse.aether.util.repository.ChainedWorkspaceReader;
+
+import com.google.common.collect.ImmutableMap;
+
+class SimpleReactorReader implements MavenWorkspaceReader {
+
```
Comment (rexhoffman) [inline takari-lifecycle-plugin/src/test/java/io/takari/maven/plugins/compile/SimpleReactorReader.java]: This feels like it should be the start of a generic maven testing library, I thought you already had one?  Assuming the use of that lib would have caused a cyclic dependency.

Comment (ifedorenko) [inline takari-lifecycle-plugin/src/test/java/io/takari/maven/plugins/compile/SimpleReactorReader.java]: Yes, agree this should go to plugin testing harness. Eventually. Maybe. But right now this is the only place where I used this testing approach and I can't tell if it's generally useful or not yet.

**Comment by rexhoffman** [comment]
Looks good, just nit picking a couple of things.   I still would like to see a generic project building lib for maven for tests, there's a lot of duplicate code around maven project test setup.

+1

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/jdt/CompilerJdt.java (hunk 1 above)
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/compile/jdt/CompilerJdt.java]: Agree this line does not belong to this change, will remove before merge.

The line is still useful because `FullCompilationStrategy` has much simpler logic, but I need to introduce "escalate to full compile" helper instead. In a future commit.
